Product filter by category performance and UI improvements.

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

This is part of #4578
<!-- Id number of the GitHub issue this PR addresses. -->

### What this PR does

The amount of line changes is rather big, but a good amount of it is also just the skeleton XML and navigation setup. I've added various comments to hopefully help explain things, but please let me know if anything is unclear.

Here is the breakdown of all the added changes:

**1. Loading improvements to product filtering.**
Speaking generally, the product filtering flow goes like this:

```
 [Product List screen] -> [Filter Items screen] -> [Filter Item Options screen]
```

Previously, list of categories is loaded on Filter Items screen. With this PR, list of categories is loaded on Filter Item Options screen, instead. The PR makes Filter Items load faster, and adds a bit of caching after fetching list of categories.

**2. Load more categories.**
Previously, the category filter item option screen only loads 100 categories. With this PR, if the user scrolls down to the end and the site has more than 100 categories, then the app will load the next 100 categories, and so on.

This also includes a change in `ProductFilterOptionListAdapter` to inherit from `ListAdapter` instead of `RecyclerView.Adapter` for better diffing. They're mostly boilerplate code, though, and the only real change is the way it does load more call.

**3. Selected category name preservation.**
Due to how the existing code is set up, some code is also needed preserve selected category option name. Selected options preservation works like this: If a user already does a filtering, then wants to do another, then the previously selected filtering options are shown on the Filter Items screen. Here we want to preserve category name because it's the data we need to show.

 Quite a bit of parts in the PR has to do with this, especially with preserving and data passing for selected category name when moving between the three screens:
- From Filter Item Options to Filter Items, the selected data doesn't need to be passed as it's already stored in the shared  `ProductFilterListViewModel`
- From Filter Items to Products List, data are sent via `ExitWithResult` event.
- From Product List screen to Filter Items screen, the same data is sent via nav argument.
- There's also some `savedState` work to preserve category name in the event of process death, both on Products List and Filter Items screens. This follows existing code that already preserves the options.

**4. UI updates.**
Finally, this add a couple of UI changes:
- Skeleton loading on the options screen when loading categories
- Circular progress loading during load more categories action.

<hr />

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

**1. Loading improvements to product filtering.**
1. Go to Products List, tap Filter. 
2. Select "Category".
3. Assuming there's active internet connection, the app will show the loading skeleton and eventually show the list of categories. Make sure "Any" is shown at the top.
4. Select a category and go back to the Filter Item screen.
5. On Filter Item screen, tap "Category" again and make sure the list of categories now loads instantly instead of having to fetch and show skeleton again.
6. Select a category, complete the filtering, and check the result on the Products List screen.

**2. Load more categories.**
1. Use a test site with more than 100 categories. The simplest way might be via WP CLI command `wp term generate product_cat`, which will generate 100 example product categories. There is also [a plugin that can help](https://te.wordpress.org/plugins/bulk-woocommerce-category-creator/).
2. Go to Products List, tap Filter. 
2. Select "Category".
3. Assuming there's active internet connection, the app will show the loading skeleton and eventually show the list of categories.
4. Scroll to the end of the list, and make sure the circular progress bar is shown at the bottom as the app tries to load more categories.
5. Wait and make sure the next set of categories show up.
6. Pick a category from that next set, complete the filtering, and check the result on the Products List screen.

**3. Selected category name preservation.**
1. Go to Products List, tap Filter. 
2. Select "Category".
3. Assuming there's active internet connection, the app will show the loading skeleton and eventually show the list of categories. Make sure "Any" is shown at the top.
4. Select a category, remember its name.
5. Complete the filtering and check the result on the Products List screen.
6. Tap "Filter" again, this will open the Filter Items screen.
7. Make sure the selected category name is listed to the next of "Category" item.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Update various code comments to mention category filter.

hafizrahman committed:
Include selected category in hasChanges check.

hafizrahman committed:
Handle different cases on how to build the category options.

See comment included in buildCategoryFilterListItemUiModel() for more details of what's being handled.

hafizrahman committed:
Move category fetching to be done only when category filter is selected.

hafizrahman committed:
Use mutablelist for productCategories, and create converter function.

This will be used to support load more categories functionality.

hafizrahman committed:
Start load more listener structure.

No actual load more functionality yet.

hafizrahman committed:
Convert adapter to ListAdapter for better diffing.

hafizrahman committed:
Add load more functionality in view model.

This also converts `productCategories` back to val, because as it turns out  `productCategoriesRepository.fetchProductCategories(loadMore = true)` returns the entire result the whole time, instead of just returning the next loaded result.

hafizrahman committed:
After load more, update category options in _filterListItems.

hafizrahman committed:
Persist selected category's name, to be used on filter item screen.

This is especially used for this use case:
1. Go to filter item screen, select "Category", then pick a category.
2. Return to filter item screen, and do the filtering.
3. This shows product list screen, with products being filtered.
4. If user taps the filter button again, this will open filter item screen again.
5. To indicate previously selected option, the filter item screen now has to show the selected category's name next to the "Category" item (instead of showing "Any").

hafizrahman committed:
Add "Any" option to the category options.

hafizrahman committed:
Add savedstate persistence for category name.

hafizrahman committed:
Improve logic on how to restore selected category name.

There are three cases:
1. User coming from product list to product filter. In this case, init() will be called, and existing selected category option is passed via `arguments.selectedProductCategoryName`, so we will use this data. At this point savedState for it will still be null, so it won't overwrite the data on loadFilters().

2. User coming from product category options to product filter. In this case init() is not called, but `selectedCategoryName` will be populated by onFilterOptionItemSelected().

3. User coming after process death. In this case init() will be called but the arguments are gone. Then in loadFilters() we try to restore `selectedCategoryName` from savedState.

hafizrahman committed:
Add savedstate persistence for category name, now in product list vm.

hafizrahman committed:
Load more fix.

hafizrahman committed:
Add category options loading skeleton view.

hafizrahman committed:
Add load more indicator; hide button when skeleton is shown.

hafizrahman committed:
Layout tweak to make sure button is not covering up last item.

The idea is mostly inspired by this:

https://stackoverflow.com/questions/56741152/recyclerview-inside-constraintlayout-start-position-at-top-of-parent

hafizrahman committed:
Only do load more action if on Category filter options screen.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterListViewModel.kt:
```diff
@@ -223,22 +262,50 @@ class ProductFilterListViewModel @Inject constructor(
+                listOf(
+                    FilterListOptionItemUiModel(
+                        selectedCategoryName!!,
+                        getFilterByProductCategory()!!,
```
This is already null-checked within isComingFromProductListWithExistingFilterByCategory() so I feel it's safe to do it like this.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterListViewModel.kt:
```diff
@@ -223,22 +262,50 @@ class ProductFilterListViewModel @Inject constructor(
+            isComingFromProductListWithExistingFilterByCategory() -> {
+                listOf(
+                    FilterListOptionItemUiModel(
+                        selectedCategoryName!!,
```
At this point since `isComingFromProductListWithExistingFilterByCategory()` is true, we can assume that `selectedCategoryName` is already filled by the navigation argument during `init`.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterOptionListFragment.kt:
```diff
@@ -88,11 +104,38 @@ class ProductFilterOptionListFragment :
+                R.layout.skeleton_product_filter_options_categories_list,
+                delayed = true
+            )
+            binding.filterOptionListBtnFrame.hide()
```
I feel it's more smooth to hide the bottom button while the categories is still being loaded.

hafizrahman committed:
Unit test fix.

hafizrahman committed:
Simplify how productCategories is populated, coming from product list.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterListViewModel.kt:
```diff
@@ -75,6 +76,8 @@ class ProductFilterListViewModel @Inject constructor(
         params
     }
 
+    private var selectedCategoryName: String? = null
```
This could've been so much simpler if it could be added into `productFilterOptions`. 

However, it is non-trivial to do because `productFilterOptions` is tightly coupled with `org.wordpress.android.fluxc.store.WCProductStore.ProductFilterOption`, and because eventually it is used for getting/fetching data as the filter parameter in `org.wordpress.android.fluxc.store.WCProductStore.getProductsByFilterOptions()`.

So it stays as its own property, even though it adds a bit more complexity when passing and preserving it.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterListViewModel.kt:
```diff
@@ -85,40 +88,86 @@ class ProductFilterListViewModel @Inject constructor(
+        if (isFromProductListWithExistingCategoryFilter()) {
+            productCategories = listOf(
+                ProductCategory(
+                    getFilterByProductCategory()!!.toLong(),
+                    selectedCategoryName!!
```
Both `getFilterByProductCategory()` and `selectedCategoryName()` is already null-checked in `isFromProductListWithExistingCategoryFilter()` so I feel it's safe to do this.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductFilterListViewModel.kt:
```diff
@@ -85,40 +88,86 @@ class ProductFilterListViewModel @Inject constructor(
-            _filterOptionListItems.value = filterItem.filterOptionListItems
+        _filterListItems.value?.let { filterListItem ->
+            val filterItem = filterListItem[selectedFilterListItemPosition]
+            if (filterItem.filterItemKey == CATEGORY) {
```
Loading categories on the Filter options screen starts here.

hafizrahman committed:
Only do load more if there's active internet.

anitaa1990 commented on WooCommerce/src/main/res/navigation/nav_graph_product_filters.xml:
```diff
@@ -24,6 +24,10 @@
             android:name="selectedProductCategory"
             app:nullable="true"
             app:argType="string" />
+        <argument
+            android:name="selectedProductCategoryName"
+            app:nullable="true"
+            app:argType="string" />
```
❓ Why do we need to have `selectedProductCategoryName` and `selectedProductCategory` separately? Both seems to be used for the same purpose (but it's possible I'm missing something here 😅 )

hafizrahman replied:
Ah, sorry for the confusion. The app needs two pieces of category-related data here:
- `selectedProductCategory` : category id, used for the filtering API call
- `selectedProductCategoryName`: category name, used for display purposes, especially when showing previously selected category on the Filter Items screen. And maybe in the future it can be used to display currently selected category on Product List screen as well.

I agree that the id part is not obvious, though! I'll rename it.

anitaa1990 replied:
Ah, that makes sense! Thanks for the detailed explanation 😄 Do you think it makes sense to pass `ProductCategory` as a whole rather than pass two separate variables here? `ProductCategory` is a parcelable class so it should be ok to pass via the navigation component. 

But that is entirely upto you and we can also do it in a separate PR :) So feel free to merge the PR if we're fine to leave it as is 👍

hafizrahman replied:
>  Do you think it makes sense to pass ProductCategory as a whole rather than pass two separate variables here

 Thanks for mentioning that! This was what I wanted to do (and was the "better solution" that I mentioned to you on Slack earlier). I explored but I found it got convoluted real quick because of the way the Product List and Product Filter List viewmodels put all selected options in a `MutableMap<ProductFilterOption, String>`. 

I thought of changing that to `MutableMap<ProductFilterOption, Any>` so I can store the entire `ProductCategory` instance in the `CATEGORY` key (while the other entries's values can still be String), but then it makes saving/checking/preserving the data more complicated due to how `Any` is used. I found null checking to be harder, for instance, but that's maybe my inexperience talking.

The `MutableMap<ProductFilterOption, String>` is also eventually used as parameter in the API call, so there will have to be added code to convert, while at the moment we can just use it as-is.

When I have time I'll be happy to revisit though. There's potential to simplify the code indeed if `ProductCategory` can be used here.

anitaa1990 reviewed (approved):
Excellent work on this @hafizrahman! The experience is so much better here <3 .

I did leave a question in the code because I am unsure of the need of having two separate category filters when navigating to different screens. 

One minor issue when testing was that the filter list screen is not scrollable when in landscape mode. It was fine because with just 3 filter options but since adding categories, it might be time to make it scrollable.

hafizrahman committed:
Clarify difference between selected category id and category name.

hafizrahman committed:
Improve checking whether to load categories or not.

Also introduces `isProductCategoriesPartiallyFilled()` and some comments to better explain what's going on.

hafizrahman committed:
Make sure filter item list is scrollable on landscape mode.

Also includes changes so that in case filter items are added in the future, the bottom button will not cover up the item.

anitaa1990 reviewed (approved):
LGTM :shipit: Thanks for working on this one @hafizrahman! 

Just a note that it would be nice to include some release notes for this before merging into `develop`.

hafizrahman commented:
> Just a note that it would be nice to include some release notes for this before merging into develop.

Thank you for the reminder! Will do it on the PR to merge the feature branch.
